### PR TITLE
Healthcheck

### PR DESCRIPTION
**Pull request by comtihon**
Add healthcheck, switch 2016->2017 in copyrigtht.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,13 +2,15 @@ PROJECT = brucke
 PROJECT_DESCRIPTION = Inter-cluster bridge of kakfa topics
 PROJECT_VERSION = $(shell cat VSN)
 
-DEPS = lager brod yamerl graphiter
+DEPS = lager brod yamerl graphiter cowboy jsone
 
 dep_brod_commit = 2.2.15
```
Comment (id) [inline Makefile]: please bump this to 2.2.16

File: Makefile [hunk 2]
```diff
@@ -2,13 +2,15 @@ PROJECT = brucke
...
 dep_yamler = git https://github.com/yakaz/yamerl.git
 dep_yamerl_commit = v0.3.2-1
 dep_graphiter = git https://github.com/klarna/graphite-erlang.git
 dep_graphiter_commit = 1.0.4
+dep_cowboy = git ssh://git@stash.int.klarna.net:7999/external/cowboy.git 2.0.0-pre.3
+dep_jsone = git ssh://git@stash.int.klarna.net:7999/external/jsone.git 1.4.1
```
Comment (id) [inline Makefile]: I think we should use regular deps here, not internal ones

**Comment by id** [comment]
When brucke fails to connect to kafka we just get 500. Should we perhaps handle this?

**Comment by id** [comment]
Please bump the version

**Comment by comtihon** [comment]
Do we need to add an ability for healthcheck to handle https?
From my point of view it is not needed, as bruck should also be installed in dmz with ports available for everyone closed.

File: Makefile (hunk 2 above)
Comment (comtihon) [reply Makefile]: ok

**Comment by id** [comment]
Don't think we need to support https here. At least for now.

**Comment by id** [comment]
One last thing: in existing files where you changed 2016 to 2017, make it `2016-2017`.

File: src/brucke_app.erl [hunk 3]
```diff
@@ -1,5 +1,5 @@
 %%%
-%%%   Copyright (c) 2016 Klarna AB
+%%%   Copyright (c) 2017 Klarna AB
```
Comment (zmstone) [inline src/brucke_app.erl]: 2017 for newly added files.
2016-2017 for modified files.
Do not change a file only to update Copyright notice years.

File: README.md [hunk 4]
```diff
@@ -57,10 +55,64 @@ Generate a release and package it into an rpm package:
...
+                    "topic": "topic_1"
+                }
+            }
+        ]
+    
+    }
+Where:  
+__routes__ is a list of `brucke_member_sup` processes.  
+__members__ is a list of `brucke_member` processes.  
+__consumers__ and __producers__ can be a list or an object with `{"status" : Error}` meaning that no
```
Comment (zmstone) [inline README.md:104]: explain Error in README?

Comment (comtihon) [reply README.md:104]: ok

File: README.md [hunk 5]
```diff
@@ -57,10 +55,64 @@ Generate a release and package it into an rpm package:
...
+        ]
+    
+    }
+Where:  
+__routes__ is a list of `brucke_member_sup` processes.  
+__members__ is a list of `brucke_member` processes.  
+__consumers__ and __producers__ can be a list or an object with `{"status" : Error}` meaning that no 
+consumers/producers can be found for this id and topic.
+
+### Custom query
```
Comment (zmstone) [inline README.md]: Query

Comment (zmstone) [inline README.md]: Add an example ?

Comment (comtihon) [reply README.md]: Why query should be upper?

File: src/brucke_config.erl [hunk 6]
```diff
@@ -1,5 +1,5 @@
 %%%
-%%%   Copyright (c) 2016 Klarna AB
+%%%   Copyright (c) 2016-2017 Klarna AB
```
Comment (zmstone) [inline src/brucke_config.erl:2]: revert.

Comment (zmstone) [reply src/brucke_config.erl:2]: never mind, no big deal.
but please keep in mind as i also commented earlier: we don't update a file only for the copyright notice years.

Comment (comtihon) [reply src/brucke_config.erl:2]: on every file I didn't change or only on this?

Comment (comtihon) [reply src/brucke_config.erl:2]: ok

File: src/brucke_health_check.erl [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        undefined -> #{ClientId => undefined};
+        _ -> #{ClientId => ok}
+      end
+    end, AllClients);
+get_all_clients(_) -> [].
+
+%% @private
+-spec get_all_routes(binary()) -> list(map()).
+get_all_routes(true) ->
+  AllRoutes = brucke_config:all_routes(),
```
Comment (zmstone) [inline src/brucke_health_check.erl]: this is the result taken from ets table.
if a route is conigured wrong in yml, bruke will start, but with the bad route discarded.
(an 'ALERT' level log is emitted at start)

We should perhaps also keep a record of the discarded routes (for example in another ets)

File: src/brucke_health_check.erl [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    true ->
+      Members = get_all_brucke_members(Route),
+      #{?STATUS => ok, ?MEMBERS => Members};
+    false ->
+      #{?STATUS => <<"dead">>, ?MEMBERS => []}
+  end.
+
+%% @private
+get_brod_data({Id, Topic}, GetFun) ->
+  case brod:get_partitions_count(Id, Topic) of
```
Comment (zmstone) [inline src/brucke_health_check.erl]: Currently, all group members are running in the same Erlang VM.
We can however potentially have two or more brucke Erlang nodes, or even different hosts.
(they should run the exact same config).
The members will join the same group and partitions are distributed between nodes.

I don't think we need such detailed per-partition status report.
A status report at group member level should be enough.
`brucke_member` is linked to ALL per-partition-assigment workers,
if any of the worker is having trouble fetching messages from upstream or
producing messages to downstream, it will crash and cause `brucke_memger` to restart.

Comment (zmstone) [reply src/brucke_health_check.erl]: In case you wonder: `brucke_member` is always started and stands by even if no assignment is received.

File: README.md [hunk 9]
```diff
@@ -136,4 +137,39 @@ Example:
         ],
         "routes": [ ]
     
+    }
+Example:
+`http://127.0.0.1:8080/health?routes_depth=party`
```
Comment (id) [inline README.md]: what's `party`?

File: src/brucke_http_handler.erl [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+init(Req, Opts) ->
+  Req2 = act(cowboy_req:path(Req), Req),
+  {ok, Req2, Opts}.
+
+
+%% @private
+act(<<"/health", _/binary>>, Req) ->
+  QS = cowboy_req:parse_qs(Req),
+  Health = brucke_health_check:get_report_data(QS),
+  reply(Req, 200, jsone:encode(Health));
```
Comment (id) [inline src/brucke_http_handler.erl]: can we have different status code if something is wrong?

Comment (comtihon) [reply src/brucke_http_handler.erl]: We need to specify criteria for wrong.

Comment (id) [reply src/brucke_http_handler.erl]: status is not "ok" for anything

Comment (comtihon) [reply src/brucke_http_handler.erl]: ok

File: src/brucke_int.hrl [hunk 11]
```diff
@@ -67,6 +67,8 @@
 -record(route, { upstream   :: {brod_client_id(), kafka_topic()}
                , downstream :: {brod_client_id(), kafka_topic()}
                , options    :: route_options()
+               , status = ok :: atom()
```
Comment (zmstone) [inline src/brucke_int.hrl]: wouldn't it be easier to create another ets for inactive routes ?

File: src/brucke_routes.erl [hunk 12]
```diff
@@ -120,6 +129,21 @@ validate_route(RawRoute0) ->
       {error, Reasons}
   end.
 
+%% @private
+mark_not_active_route(Route, Status, Reason) ->
+  MayBeId = proplists:get_value(upstream_client, Route),
```
Comment (zmstone) [inline src/brucke_routes.erl]: not_active -> inactive
MayBe ->  Maybe

File: src/brucke_int.hrl (hunk 11 above)
Comment (comtihon) [reply src/brucke_int.hrl]: no

File: src/brucke_routes.erl [hunk 13]
```diff
@@ -52,10 +52,11 @@ init(Other) ->
   erlang:exit(bad_routes_config).
 
 %% @doc Lookup in config cache for the message routing destination.
--spec lookup(brod_client_id(), kafka_topic()) -> route() | false.
-lookup(UpstreamClientId, UpstreamTopic) ->
+-spec lookup_ok(brod_client_id(), kafka_topic()) -> route() | false.
+lookup_ok(UpstreamClientId, UpstreamTopic) ->
   case ets:lookup(?ETS, {UpstreamClientId, UpstreamTopic}) of
     []      -> false;
+    [#route{status = NotOk}] when NotOk /= ok -> false;
```
Comment (zmstone) [inline src/brucke_routes.erl]: why not match `ok` instead  of NotOk with a guard ?

Comment (comtihon) [reply src/brucke_routes.erl]:
```
[Route = #route{status = ok}] -> Route;
_ -> false
```
?

File: src/brucke_health_check.erl [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+-module(brucke_health_check).
+
+-define(UPSTREAM, <<"upstream">>).
+-define(DOWNSTREAM, <<"downstream">>).
+-define(PRODUCERS, <<"producers">>).
+-define(CONSUMERS, <<"consumers">>).
+-define(STATUS, <<"status">>).
+-define(MEMBERS, <<"members">>).
+-define(TOPIC, <<"topic">>).
+-define(ID, <<"id">>).
```
Comment (zmstone) [inline src/brucke_health_check.erl]: please change to `<<"client_id">>`

Comment (zmstone) [reply src/brucke_health_check.erl]: and more clear to rename the macro to CLIENT_ID also.

Comment (comtihon) [reply src/brucke_health_check.erl]: ok

File: src/brucke_health_check.erl [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    {_, RoutePid, _, _} when is_pid(RoutePid) ->  % ok / dead (but pid found in sup tree)
+      check_process(RoutePid, Status);
+    {_, Status, _, _} ->  % undefined / restarting
+      {#{?STATUS => Status}, false};
+    false ->  % not in tree
+      {#{?STATUS => <<"not under brucke_sup">>}, false}
+  end.
+
+%% @private
+check_process(Route, Status) ->
```
Comment (zmstone) [inline src/brucke_health_check.erl]: add spec?

Comment (comtihon) [reply src/brucke_health_check.erl]: Do we really need spec on private function?
It gets pid and boolean from check_route some lines above and returns tuple, as we can see.

File: src/brucke_health_check.erl [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+  case is_process_alive(Route) of
```
Comment (zmstone) [inline src/brucke_health_check.erl]: there is a race here which can lead to unnecessary 500.
perhaps don't check process live or not, do this instead:
```
try 
  Members = get_all_brucke_members(RoutSupPid),
  {#{?STATUS => ok, ?MEMBERS => Members}, Status}
catch exit : {noproc, _} -> %% maybe some other exit reason too?
  {#{?STATUS => <<"dead">>, ?MEMBERS => []}, false}
end
```

File: README.md [hunk 17]
```diff
@@ -57,10 +55,73 @@ Generate a release and package it into an rpm package:
...
+    
+        "clients": [
+            {
+                "brod_client_1": "ok"
+            }
+        ],
+        "routes": [
+            {
+                "downstream": {
+                    "id": "brod_client_1",
```
Comment (zmstone) [inline README.md]: client_id

File: src/brucke_health_check.erl [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+      {[#{?UPSTREAM => #{?CLIENT_ID => Id, ?TOPIC => Topic}, ?STATUS => Reason} | Acc], false}
+    end, {Working, Status}, DiscardedRoutes).
+
+%% @private
+check_route(Upstream, Routes, Status) ->
+  case lists:keyfind(Upstream, 1, Routes) of
+    {_, RoutePid, _, _} when is_pid(RoutePid) ->  % ok / dead (but pid found in sup tree)
+      check_process(RoutePid, Status);
+    {_, Status, _, _} ->  % undefined / restarting
+      {#{?STATUS => Status}, false};
```
Comment (zmstone) [inline src/brucke_health_check.erl]: here and line 101, add `members => []`
maybe add a function `make_route_satus(Status, Members)` to ensure unified schema in JSON response.

File: src/brucke_health_check.erl [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+    {#{?STATUS => <<"dead">>, ?MEMBERS => []}, false}
+  end.
+
+%% @private
+get_all_brucke_members(RoutePid) ->
+  Members = brucke_sup:get_children(RoutePid),
+  lists:foldl(
+    fun
+      ({Id, ChildStatus, _, _}, Acc) when is_atom(ChildStatus) ->  % undefined/restarting
+        Acc#{<<<<"brucke_member">>/binary, (integer_to_binary(Id))/binary>> => ChildStatus};
```
Comment (zmstone) [inline src/brucke_health_check.erl]: we don't need this "brucke_member" prefix.

File: README.md (hunk 17 above)
Comment (zmstone) [reply README.md]: please update example

**Comment by zmstone** [comment]
looks good, i'll play with it a bit before approve.

File: src/brucke_http_handler.erl [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+act(_, Req) ->
+  reply(Req, 404, jsone:encode(#{<<"error">> => <<"Not found">>})).
+
+%% @private
+-spec reply(cowboy_req:req(), integer(), binary()) -> cowboy_req:req().
+reply(Req, Code, Data) ->
+  cowboy_req:reply(Code, ?JSON, Data, Req).
+
+%% @private
+status_to_code(true) -> 100;
```
Comment (id) [inline src/brucke_http_handler.erl]: 200

**Comment by zmstone** [comment]
the code was good (except for one small bug).
but it was hard to follow when missing the specs.
specs like `map()` and `list()` are as good as no specs at all.

**Comment by comtihon** [comment]
What bug?

**Comment by zmstone** [comment]
this bug: https://github.com/klarna/brucke/pull/9

**Review by id — APPROVED** [review]

**Review by zmstone — APPROVED** [review]
